mlx/backend/metal/qrf.cpp

```diff
@@ -0,0 +1,696 @@
+  float* y_data = Y.data<float>();
+  float* betas_data = betas.data<float>();
+  float* yp_data = Yp.data<float>();
+  float* wy_data = WY.data<float>();
+  float* qwy_data = QWY.data<float>();
+  float* wp_data = Wp.data<float>();
+
+  auto& stream = q.primitive().stream();
+  auto& device = metal::device(stream.device);
+
+  for (int jj = 0; jj < R; jj++) {
+    const int col = startc + jj;
+
+    if constexpr (!RUN_ON_CPU_FOR_DEBUG) {
+      auto kernel = device.get_kernel("qrf_compute_v");
+      auto command_buffer = device.new_command_buffer(stream.index);
+
+      auto compute_encoder =
+          metal::CommandEncoder(command_buffer->computeCommandEncoder());
+      compute_encoder->setComputePipelineState(kernel);
+
+      // todo: pass float* directly to support stacking.
+      compute_encoder.set_input_array(a, 0);
+      compute_encoder.set_input_array(Y, 1);
+      compute_encoder.set_input_array(betas, 2);
+      compute_encoder->setBytes(&jj, sizeof(jj), 3);
+      compute_encoder->setBytes(&startc, sizeof(startc), 4);
+      compute_encoder->setBytes(&m, sizeof(m), 5);
+
+      // Specify the grid dimension and use the largest possible threads groups,
+      // see
+      // https://developer.apple.com/documentation/metal/compute_passes/calculating_threadgroup_and_grid_sizes
+      const MTL::Size threads_per_grid(m, 1, 1);
+      // todo: share threadgroup size with the kernel shared memory buffer.
+      const MTL::Size threads_per_threadgroup(8, 1, 1);
+      compute_encoder->dispatchThreads(
+          threads_per_grid, threads_per_threadgroup);
+      compute_encoder->endEncoding();
+
+      device.commit_command_buffer(stream.index);
+    } else {
+      // Accumulate norm(x) where x is the part of the current column on and
+      // below the diagonal.
+      float sq_norm_xc = 0;
+
+      // NOTE: only the elements on and below the diagonal.
+      for (int i = col; i < m; i++) {
+        sq_norm_xc += std::pow(a_data[colmajor_idx(i, col, m)], 2);
+      }
+
+      float norm_xc = std::pow(sq_norm_xc, 0.5);
+
+      // Build Householder vector v (stored in the jj-th column of Y).
+      float* v = y_data + colmajor_idx(0, jj, m);
+      float sq_norm_v = 0;
+      for (int i = 0; i < m; i++) {
+        if (i < col) {
+          v[i] = 0;
+        } else if (i == col) {
+          const float old_val = a_data[colmajor_idx(i, i, m)];
+          const float new_val =
+              old_val >= 0 ? old_val + norm_xc : old_val - norm_xc;
+          v[i] = new_val;
+        } else {
+          v[i] = a_data[colmajor_idx(i, col, m)];
+        }
+        sq_norm_v += std::pow(v[i], 2);
+      }
+
+      const float beta = 2.0f / sq_norm_v;
+      betas_data[jj] = beta;
+    }
+
+    // Apply reflection to the rest of the current block.
+    if constexpr (!RUN_ON_CPU_FOR_DEBUG) {
+      auto kernel = device.get_kernel("qrf_reflect_current_block");
+      auto command_buffer = device.new_command_buffer(stream.index);
```

Comment:
You don't need to get the command buffer here. Instead get the already open command encoder:

`d.get_command_encoder(stream.index)`